komodo/yaml_file_type.py

```diff
@@ -38,16 +134,506 @@ class ManifestFile(YamlFile):
+            )
+            Package.validate_package_maintainer_with_errors(
+                package_name,
+                version,
+                version_metadata.get("maintainer"),
+                errors,
+            )
+            for (
+                package_property,
+                package_property_value,
+            ) in version_metadata.items():
+                self.validate_package_properties(
+                    package_name,
+                    version,
+                    package_property,
+                    package_property_value,
+                    errors,
+                )
+
+    def validate_package_properties(
+        self,
+        package_name: str,
+        package_version: str,
+        package_property: str,
+        package_property_value: str,
+        errors: List[str],
+    ):
+        PRE_CHECKED_PROPERTIES = ["make", "maintainer"]
+        if package_property in PRE_CHECKED_PROPERTIES:
+            return
+        if package_property == "depends":
+            if not isinstance(package_property_value, list):
+                errors.append(
+                    f"Dependencies for package {package_name} has"
+                    f" invalid type {package_property_value}"
+                )
+                return
+            for dependency in package_property_value:
+                if not isinstance(dependency, str):
+                    errors.append(
+                        f"Package {package_name} version {package_version} has"
+                        f" invalid dependency type({dependency})"
+                    )
+                    continue
+                if dependency not in self.yml.keys():
+                    errors.append(
+                        f"Dependency '{dependency}' not found for"
+                        f" package '{package_name}'"
+                    )
+        else:
+            try:
+                Package.validate_package_property_type(
+                    package_name,
+                    package_version,
+                    package_property,
+                    package_property_value,
+                )
+            except (ValueError, TypeError) as e:
+                errors.append(str(e))
+
+
+class UpgradeProposalsFile(YamlFile):
+    """
+    Return the data from 'upgrade_proposals' YAML, but validate it first.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.yml = None
+
+    def __call__(self, value: str) -> Dict[str, Dict[str, str]]:
+        yml = super().__call__(value)
+        self.validate_upgrade_proposals_file(yml)
+        self.yml = yml
+        return self
+
+    def from_yaml_string(self, value):
+        yaml = YAML()
+        yml = yaml.load(value)
+        self.validate_upgrade_proposals_file(yml)
+        self.yml = yml
+        return self
+
+    def validate_upgrade_key(self, upgrade_key: str) -> None:
+        assert (
```

Comment:
Do you want to throw an `AssertionError` here, not something ala `KomodoError`?

Reply:
This validates that the wanted upgrade_key is in the file, and should exit if it is not present (or invalid?). We could throw KomodoException instead, but I dont think it would make a difference. Other places I use KomodoError if I want to catch the error and extract the message.